pinot-core/src/main/java/org/apache/pinot/core/data/IndexedTable.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.pinot.common.request.AggregationInfo;
+import org.apache.pinot.common.request.SelectionSort;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.EqualityUtils;
+
+
+/**
+ * Table of data records to pass data between operators
+ */
+public interface IndexedTable {
+
+  void init(@Nonnull DataSchema dataSchema, List<AggregationInfo> aggregationInfos, List<SelectionSort> orderBy,
+      int maxCapacity);
+
+  boolean upsert(@Nonnull TableRecord record);
+
+  boolean merge(@Nonnull IndexedTable table);
+
+  int size();
+
+  Iterator<TableRecord> iterator();
+
+  boolean sort();
```

Comment:
let's not expose this for now. Add a finish method and let the implementation figure out when to sort the data

Reply:
What should be happening in the finish method? were you thinking the finish will call sort? or should we let sort be an external service. I think latter is better

Reply:
Its up to the implementation. for e.g. finish can call sort if there is some order by otherwise it can be unordered.